gemini/sample-apps/e2e-gen-ai-app-starter-pack/app/utils/tracing.py

```diff
@@ -0,0 +1,150 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+from typing import Any, Optional, Sequence
+
+from google.cloud import logging as google_cloud_logging
+from google.cloud import storage
+from opentelemetry.exporter.cloud_trace import CloudTraceSpanExporter
+from opentelemetry.sdk.trace import ReadableSpan
+from opentelemetry.sdk.trace.export import SpanExportResult
+
+
+class CloudTraceLoggingSpanExporter(CloudTraceSpanExporter):
+    """
+    An extended version of CloudTraceSpanExporter that logs span data to Google Cloud Logging
+    and handles large attribute values by storing them in Google Cloud Storage.
+
+    This class helps bypass the 256 character limit of Cloud Trace for attribute values
+    by leveraging Cloud Logging (which has a 256KB limit) and Cloud Storage for larger payloads.
+    """
+
+    def __init__(
+        self,
+        logging_client: Optional[google_cloud_logging.Client] = None,
+        storage_client: Optional[storage.Client] = None,
+        bucket_name: Optional[str] = None,
+        debug: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        """
+        Initialize the exporter with Google Cloud clients and configuration.
+
+        :param logging_client: Google Cloud Logging client
+        :param storage_client: Google Cloud Storage client
+        :param bucket_name: Name of the GCS bucket to store large payloads
+        :param debug: Enable debug mode for additional logging
+        :param kwargs: Additional arguments to pass to the parent class
+        """
+        super().__init__(**kwargs)
+        self.debug = debug
+        self.logging_client = logging_client or google_cloud_logging.Client(
+            project=self.project_id
+        )
+        self.logger = self.logging_client.logger(__name__)
+        self.storage_client = storage_client or storage.Client(project=self.project_id)
+        self.bucket_name = bucket_name or f"{self.project_id}-logs-data"
+        self._ensure_bucket_exists()
+        self.bucket = self.storage_client.bucket(self.bucket_name)
+
+    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
+        """
+        Export the spans to Google Cloud Logging and Cloud Trace.
+
+        :param spans: A sequence of spans to export
+        :return: The result of the export operation
+        """
+        for span in spans:
+            span_context = span.get_span_context()
+            trace_id = format(span_context.trace_id, "x")
+            span_id = format(span_context.span_id, "x")
+            span_dict = json.loads(span.to_json())
+
+            span_dict["trace"] = f"projects/{self.project_id}/traces/{trace_id}"
```

Comment:
There should be a method on the client to create this path. Something like `client.trace_path()`